dds/DCPS/transport/tcp/TcpReconnectTask.cpp

```diff
@@ -6,48 +6,93 @@
+#include "dds/DCPS/transport/tcp/TcpConnection.h"
+#include "dds/DCPS/transport/tcp/TcpTransport.h"
 #include "dds/DCPS/transport/framework/EntryExit.h"
 
+#include "ace/Reverse_Lock_T.h"
+
 OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 
-OpenDDS::DCPS::TcpReconnectTask::TcpReconnectTask(
-  OpenDDS::DCPS::TcpConnection* connection)
-  : connection_(connection)
+namespace OpenDDS {
+namespace DCPS {
+
+TcpReconnectTask::TcpReconnectTask() : connection_(), mutex_(), cv_(mutex_), shutdown_(false)
 {
-  DBG_ENTRY_LVL("TcpReconnectTask","TcpReconnectTask",6);
+  DBG_ENTRY_LVL("TcpReconnectTask", "TcpReconnectTask", 6);
   if (open()) {
     ACE_ERROR((LM_ERROR,
                ACE_TEXT("(%P|%t) ERROR: Reconnect task failed to open : %p\n"),
                ACE_TEXT("open")));
   }
 }
 
-OpenDDS::DCPS::TcpReconnectTask::~TcpReconnectTask()
+TcpReconnectTask::~TcpReconnectTask()
+{
+  DBG_ENTRY_LVL("TcpReconnectTask", "~TcpReconnectTask", 6);
+  shutdown();
+  wait();
+}
+
+void TcpReconnectTask::reconnect(TcpConnection_rch connection)
+{
+  DBG_ENTRY_LVL("TcpReconnectTask", "reconnect", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  if (!shutdown_ && !connection_) {
+    connection_ = connection;
+    activate(1);
+  }
+}
+
+bool TcpReconnectTask::active() {
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  return connection_;
+}
+
+void TcpReconnectTask::shutdown() {
+  DBG_ENTRY_LVL("TcpReconnectTask", "shutdown", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  shutdown_ = true;
+}
+
+void TcpReconnectTask::wait_complete()
 {
-  DBG_ENTRY_LVL("TcpReconnectTask","~TcpReconnectTask",6);
+  DBG_ENTRY_LVL("TcpReconnectTask", "wait_complete", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  while (connection_) {
+    cv_.wait();
+  }
 }
 
-void OpenDDS::DCPS::TcpReconnectTask::execute(ReconnectOpType& op)
+int TcpReconnectTask::svc()
 {
-  DBG_ENTRY_LVL("TcpReconnectTask","execute",6);
-
-  // Ignore all signals to avoid
-  //     ERROR: <something descriptive> Interrupted system call
-  // The main thread will handle signals.
-  sigset_t set;
-  ACE_OS::sigfillset(&set);
-  ACE_OS::thr_sigsetmask(SIG_SETMASK, &set, NULL);
-
-  if (op == DO_RECONNECT) {
-    if (this->connection_->reconnect() == -1) {
-      this->connection_->tear_link();
+  DBG_ENTRY_LVL("TcpReconnectTask", "svc", 6);
+
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, 0);
+
+  // The order here matters for destruction if we're the last reference holders by the end
+  TcpTransport_rch impl(connection_->impl());
+  TcpConnection_rch connection(connection_);
+
+  ACE_Reverse_Lock<ACE_Thread_Mutex> rev_lock(mutex_);
+
+  if (connection) {
+    ACE_GUARD_RETURN(ACE_Reverse_Lock<ACE_Thread_Mutex>, rev_guard, rev_lock, 0);
+    if (connection->reconnect() == -1) {
+      connection->tear_link();
     }
+  }
+
+  connection_.reset();
+  cv_.broadcast();
 
-  } else
-    ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: TcpReconnectTask::svc unknown operation %d\n", op));
+  g.release();
```

Comment:
Is this needed?

Reply:
Yeah, if we happen to be the last reference holder for TcpConnection, we need to allow it to call back in to TcpReconnectTask::shutdown().